core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -611,6 +611,8 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
         AppInfoParser.unregisterAppInfo(jmxPrefix, config.brokerId.toString)
         shutdownLatch.countDown()
         info("shut down completed")
+      } else {
+        isShuttingDown.set(false)
```

Comment:
This fixes the case where we `countDown()` is invoked between the `compareAndSet` and the `if` condition. What if `countDown()` is not invoked between those two calls though? It seems that we would do a second `shutdown()`?

I think it may be reason to reason about this if we use synchronization to ensure that the latch and `isShuttingDown` are updated together in atomic fashion.